### PR TITLE
[refactored] Implement dive summary with a model

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

As discussed in a previous PR, this implements a model that represents the dive summaries. The model features an arbitrary number of columns that can be calculated independently. The annoying part is that QML's ListView works via Roles, wheres QtWidget's TableView via columns. Therefore, this model does some crazy things to translate between the two.

This is only a proof-of-concept for comparison purpose with a more object-based approach. Notably, the display is not properly implemented, as I have no idea how to do that. Moreover, the commit-history is garbage, as this is based on the work of the previous, unmerged PR.

### CODE REVIEW

**Comment by Subsurface-divelog** [comment]
My general comment is the same as what I said earlier.
The Summary page is a quick hack for a specific feature.
I really hate the idea of investing this much work into it.
Unless you want to turn this into a broader Statistics module that is used both on desktop and mobile.

Heck, it's your time. And I'm of course willing to look into the QML side if you convince me that this is the way to go...

**Comment by dirkhh** [comment]
Sorry for the comments as @Subsurface-divelog earlier - I was logged into that account as I was doing some GitHub maintenance...

While it always weirds me out to push into someone else's branch, I apparently am able to do so, and it seemed the best way to continue here. This cleans up the UI and creates something that I find fairly reasonable.

So yeah, I'm happy if you want to clean up the history and create a "real" PR :-)

**Comment by dirkhh** [comment]
OK, builds are fixed, UI is "ok-ish". Over to you @bstoeger for history cleanup and other work that you may still want to do...
I'm hoping to do mobile betas this weekend, but nothing is pushing me to do that, so if this isn't done today I can simply delay the betas. I'm sure there are many other areas that are worth poking at.

**Comment by dirkhh** [comment]
did you want me to clean up the history or are you planning to do that (I notice you adding commits at the end... nothing wrong with that, just checking intentions)

**Comment by bstoeger** [comment]
> did you want me to clean up the history or are you planning to do that (I notice you adding commits at the end... nothing wrong with that, just checking intentions)

I'm going to do that - no worries. I just wanted to check your opinion on a few more UI and reorganization changes beforehand.

BTW: Is there a general rule when combining commits of two authors, like do you add both signed-off-lines?

**Comment by bstoeger** [comment]
> My general comment is the same as what I said earlier.
The Summary page is a quick hack for a specific feature.
I really hate the idea of investing this much work into it.

I understand your reservation and of course there are much more interesting / important things. For me it's not so much about the feature itself, but to gain some experience in how far desktop and mobile can use shared code / infrastructure. That's all uncharted territory for me. Sadly, I'm getting more pessimistic in that respect - QML and QWidgets are surprisingly different in their behavior. Even the model-abstraction is not really useful: roles vs. columns, but also the way in which section headers are realized. [You must be extremely fed-up with my QML rants by now, but seriously: Why did they choose Qt's object-system, which is completely unsuited for a Javascript-style framework and then minimize code reusability, which appears to be the best argument for that choice. So weird!]

A second point where I have trouble is that, at the end of the day, the UIs are quite different. For example I don't see how the mobile-UI of this feature would be very useful on desktop. Perhaps I'm missing something, but is there are reason for having two columns?

> Unless you want to turn this into a broader Statistics module that is used both on desktop and mobile.

Sadly, I don't think the UI and model as they are would be particularly useful for a comprehensive statistics model. Are there any mock-ups for such a module? I think I am simply too uninspired. I would just present start/stop date and output things like what we already have on desktop. And yes, then perhaps yearly / monthly / weekly stats...? Percentages? Pie charts? Sounds like *a lot* of work.

> Heck, it's your time. And I'm of course willing to look into the QML side if you convince me that this is the way to go...

Originally, I wanted to make two PRs, one with the model and one with structured QObjects with properties for comparison. However, given that you spent quite a lot of time (I presume) on styling this, that appears to me the best argument to go forward with this.

**Comment by dirkhh** [comment]
> > did you want me to clean up the history or are you planning to do that (I notice you adding commits at the end... nothing wrong with that, just checking intentions)
> 
> I'm going to do that - no worries. I just wanted to check your opinion on a few more UI and reorganization changes beforehand.
> 
> BTW: Is there a general rule when combining commits of two authors, like do you add both signed-off-lines?

There is a protocol for designating multiple authors, but that's not broadly used and many tools don't really support it. I'd simply go and set the author to be the person who did the bigger part of the work and have both signed-off-by lines. If it seems important, do the whole "Co-authored-by:" thing before the SOB lines

**Comment by dirkhh** [comment]
> I understand your reservation and of course there are much more interesting / important things. For me it's not so much about the feature itself, but to gain some experience in how far desktop and mobile can use shared code / infrastructure. That's all uncharted territory for me. Sadly, I'm getting more pessimistic in that respect - QML and QWidgets are surprisingly different in their behavior. Even the model-abstraction is not really useful: roles vs. columns, but also the way in which section headers are realized. [You must be extremely fed-up with my QML rants by now, but seriously: Why did they choose Qt's object-system, which is completely unsuited for a Javascript-style framework and then minimize code reusability, which appears to be the best argument for that choice. So weird!]

Yes, I may have heard you once say something negative about QML already. Maybe twice. :-)
And I share your frustration about the course they went.
What most people here will not know is that I was around when QML was created and pushed and when some of these decisions were made by Nokia (who then owned Qt after acquiring TrollTech). Intel had a collaboration with them around this technology stack. And as often happens, decisions were made based on what could be implemented to meet certain timelines...

> A second point where I have trouble is that, at the end of the day, the UIs are quite different. For example I don't see how the mobile-UI of this feature would be very useful on desktop. Perhaps I'm missing something, but is there are reason for having two columns?

The two columns were the brain child of Jan - that's something that he has heard dive centers ask for. Total dive experience and "freshness" of that experience, i.e. what have you done in the last month/quarter/year. And I have every intention of fixing the "multiples of 30 day thing" before we do a release.

> > Unless you want to turn this into a broader Statistics module that is used both on desktop and mobile.
> 
> Sadly, I don't think the UI and model as they are would be particularly useful for a comprehensive statistics model. Are there any mock-ups for such a module? I think I am simply too uninspired. I would just present start/stop date and output things like what we already have on desktop. And yes, then perhaps yearly / monthly / weekly stats...? Percentages? Pie charts? Sounds like _a lot_ of work.

It is. Which is why several people have shown interest and even implemented part of this, but no one is actually working on it, it seems. See @atdotde 's PR #2523 

> > Heck, it's your time. And I'm of course willing to look into the QML side if you convince me that this is the way to go...
> 
> Originally, I wanted to make two PRs, one with the model and one with structured QObjects with properties for comparison. However, given that you spent quite a lot of time (I presume) on styling this, that appears to me the best argument to go forward with this.

Yeah, I undercommunicated this. I looked at the two branches you showed me and figured "heck, if we do it, then let's use a model and create at least some flexibility and even the basis of maybe taking this to a bit more of a statistics framework". That's why I cleaned up the UI (which was maybe a total of 2h of work...)

And as I indicated - I'm ok with merging a cleaned-up version of this. What's there right now is dramatically better than what we had before.

File: mobile-widgets/qml/DiveSummary.qml [hunk 1]
```diff
@@ -110,7 +110,7 @@ Kirigami.ScrollablePage {
 						id: headerLabel
 						verticalAlignment: Qt.AlignVCenter
 						colorBackground: parent.color
-						text: header !== undefined ? header : ""
+						text: header !== undefined ? "   " + header : ""
```
Comment (dirkhh) [inline mobile-widgets/qml/DiveSummary.qml]: why is there the hard white space? Are you trying to create some margin on the left?
Happy to do that in a more QML "native way" :-)

File: qt-models/divesummarymodel.cpp [hunk 2]
```diff
@@ -61,6 +62,14 @@ QVariant DiveSummaryModel::data(const QModelIndex &index, int role) const
 		return dataDisplay(row, 0);
 	case COLUMN1_ROLE:
 		return dataDisplay(row, 1);
+	case SECTION_ROLE:
+		switch (row) {
+		case DIVES ... PLANS: return tr("Number of dives");
+		case TIME ... TIME_AVG: return tr("Time");
+		case DEPTH_MAX ... DEPTH_AVG: return tr("Depth");
+		case SAC_MIN ... SAC_AVG: return tr("Depth");
```
Comment (dirkhh) [inline qt-models/divesummarymodel.cpp]: He. Maybe "SAC"? Just a guess

**Comment by dirkhh** [comment]
Added a few comments. I think your changes make sense. They now __really__ scream for some refactoring and squashing of commits %-P

**Comment by dirkhh** [comment]
ok, found time to build it and try it. Yeah, the UI once again deserves some fine tuning :-)

(see how delicate I can be in my comments?)

But the direction is fine.

Do you want me to add to this branch as is or should I wait for your rebase?

**Comment by bstoeger** [comment]
> Do you want me to add to this branch as is or should I wait for your rebase?

Yes, please add to this branch - as I'll not be able to rebase tonight.

**Comment by bstoeger** [comment]
Attn: Still working on refactor...

**Comment by bstoeger** [comment]
Alright - that's my take at refactoring the commits.

@dirkhh: I squashed a few commits of yours, please let me know if this is OK.

Github does not present the commits in logical order - very confusing.

Is there a git functionality that can checkout every commit in a series and build it? Probably I made some error here or there. When I amended a commit, I compiled, but I cannot guarantee that I didn't introduce side effects that would make other commits fail.

**Comment by dirkhh** [comment]
I have written little scripts that do the "compile every commit" thing. And often when refactoring things, I just do that manually as I walk through the commits one last time, cleaning up the commit messages, etc.

I looked at this and really like what you did with the refactoring. Those are reasonable, logical commits. Well done.

Some gentle testing doesn't reveal any major issues. So I merged this locally (because I was testing it, anyway) and pushed it out.

**Pull request closed without merging**